apps/docusaurus/docs/guides/randomness.md

```diff
@@ -0,0 +1,314 @@
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+where:
+
+- `let winner_idx = aptos_framework::randomness::u64_range(0, n);` is the a randomness API call that returns a u64 integer in range `[0, n)` uniformly at random.
+- `#[randomness]` is a required attribute for the execution engine to perform some safety check and enable the API for the transaction.
+  - Calling randomness API without this attribute on top of the entry function will cause the transaction to abort.
+  - It also means randomness API calls are supported only in transactions that call an entry function.
+
+## How to use Aptos randomness API
+
+### Currently supported network
+
+Randomness APIs are enabled on devnet and testnet, and landing soon on mainnet.
+
+### Build aptos-cli from latest source
+
+Some changes have been made to aptos-cli to support compiling a randomness consuming contract.
+While the changes haven't been included in an official aptos-cli release (but coming soon),
+you can build aptos-cli from source to support your randomness-dependent contract development.
+
+```shell
+git clone https://github.com/aptos-labs/aptos-core
+cd aptos-core
+./scripts/dev_setup.sh -b
+sudo apt-get install libudev-dev
+. ~/.profile
+cargo build -p aptos
+alias aptos=target/debug/aptos
+```
+
+### Identify randomness-dependent entry functions and make them compliant
+
+For safety, randomness API calls are allowed at runtime only when the transaction invokes an entry function that is:
+
+- private, and
+- annotated with `#[randomness]`.
+
+It's now a good time to think about what user actions need randomness API, write them down, and make sure they are private and have the right attribute, as shown in the example below.
+
+```rust
+module module_owner::lottery {
+    // ...
+
+    #[randomness]
+    entry fun decide_winner() {
+        // ...
+    }
+}
+```
+
+### Call the API
+
+The APIs are public functions under `0x1::randomness` and can be referenced directly, as demonstrated in the lottery example above.
+
+```rust
+module module_owner::lottery {
+    // ...
+
+    #[randomness]
+    entry fun decide_winner() {
+        // ...
+        let winner_idx = aptos_framework::randomness::u64_range(0, n);
+        lottery_state.winner_idx = std::option::some(winner_idx);
+    }
+}
+```
+
+The above example uses function `u64_range()` but many other basic types are also supported.
+Here's a quick overview of all the API, where `T` can be one of `u8, u16, u32, u64, u128, u256`.
+
+```
+/// Generates an T uniformly at random.
+fun T_integer(): T
+
+/// Generates a number `[min_incl, max_excl)` uniformly at random.
+fun T_range(min_incl, ): T
+
+/// Generates a sequence of bytes uniformly at random
+fun bytes(n: u64): vector<u8>;
+
+/// Generate a permutation of `[0, 1, ..., n-1]` uniformly at random.
+/// If n is 0, returns the empty vector.
+fun permutation(n: u64): vector<u64>;
+```
+
+### Security warning: randomness API calls in non-entry functions
```

Comment:
can we keep a single `Security` as ### and the specific paragraphs as #### ?

Reply:
fixed